File: scripts/package_envelope.py
"""MP05 envelope primitives; manifest and engine validation are separate."""
import hashlib

PREFIX = b"MP05MicroCard signed package v5\0"
HEADER = len(PREFIX) + 8
OVERHEAD = HEADER + 32 + 65 + 64
MAX_PACKAGE = 16384


def signing_prefix(manifest, image, key):
    if len(key) != 65 or key[0] != 4:
        raise ValueError("uncompressed SEC1 key required")
    if OVERHEAD + len(manifest) + len(image) > MAX_PACKAGE:
        raise ValueError("package exceeds quota")
    return (PREFIX + len(manifest).to_bytes(4, "little") + len(image).to_bytes(4, "little")
            + manifest + hashlib.sha256(image).digest() + key)


def create(manifest, image, key, sign):
    prefix = signing_prefix(manifest, image, key)
    signature = sign(prefix)
    if len(signature) != 64:
        raise ValueError("P1363 signature required")
    return prefix + signature + image


def verify(raw, verify_signature):
    if not OVERHEAD <= len(raw) <= MAX_PACKAGE or not raw.startswith(PREFIX):
        raise ValueError("invalid MP05 envelope")
    manifest_length = int.from_bytes(raw[len(PREFIX):len(PREFIX) + 4], "little")
    image_length = int.from_bytes(raw[len(PREFIX) + 4:HEADER], "little")
    if OVERHEAD + manifest_length + image_length != len(raw):
        raise ValueError("invalid package length")
    digest_start = HEADER + manifest_length
    key_start = digest_start + 32
    signature_start = key_start + 65
    image_start = signature_start + 64
    key = raw[key_start:signature_start]
    signature = raw[signature_start:image_start]
    order = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
    r, s = int.from_bytes(signature[:32], "big"), int.from_bytes(signature[32:], "big")
    if key[0] != 4 or not 0 < r < order or not 0 < s <= order // 2:
        raise ValueError("invalid key or signature shape")
    if not verify_signature(key, raw[:signature_start], signature):
        raise ValueError("invalid signature")
    image = raw[image_start:]
    if hashlib.sha256(image).digest() != raw[digest_start:key_start]:
        raise ValueError("image digest mismatch")
    return raw[HEADER:digest_start], image, key

File: managed/MicroCard.Build/PackageEnvelope.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace MicroCard.Build;

/// <summary>MP05 authentication only; manifest and engine validation remain separate.</summary>
internal static class PackageEnvelope
{
    static ReadOnlySpan<byte> Prefix => "MP05MicroCard signed package v5\0"u8;
    public static int HeaderBytes => Prefix.Length + 8;
    public static int OverheadBytes => HeaderBytes + 32 + 65 + 64;

    public static byte[] Create(byte[] manifest, byte[] image, byte[] seed)
    {
        if ((long)OverheadBytes + manifest.Length + image.Length > 16384) throw new InvalidDataException("Package exceeds quota");
        using var output = new MemoryStream(); using var writer = new BinaryWriter(output);
        writer.Write(Prefix); writer.Write((uint)manifest.Length); writer.Write((uint)image.Length);
        writer.Write(manifest); writer.Write(SHA256.HashData(image)); writer.Write(PackageSignatures.PublicKey(seed));
        writer.Write(PackageSignatures.Sign(seed, output.ToArray())); writer.Write(image);
        return output.ToArray();
    }

    public static (byte[] Manifest, byte[] Image, byte[] Key) Verify(byte[] raw)
    {
        if (raw.Length < OverheadBytes || raw.Length > 16384 || !raw.AsSpan(0, Prefix.Length).SequenceEqual(Prefix))
            throw new InvalidDataException("Invalid MP05 envelope");
        int manifestLength = checked((int)BinaryPrimitives.ReadUInt32LittleEndian(raw.AsSpan(Prefix.Length, 4)));
        int imageLength = checked((int)BinaryPrimitives.ReadUInt32LittleEndian(raw.AsSpan(Prefix.Length + 4, 4)));
        if ((long)OverheadBytes + manifestLength + imageLength != raw.Length) throw new InvalidDataException("Invalid package length");
        int digestStart = HeaderBytes + manifestLength;
        int keyStart = digestStart + 32, signatureStart = keyStart + 65, imageStart = signatureStart + 64;
        byte[] key = raw[keyStart..signatureStart];
        if (!PackageSignatures.Verify(key, raw[..signatureStart], raw[signatureStart..imageStart]))
            throw new InvalidDataException("Invalid signature");
        byte[] image = raw[imageStart..];
        if (!CryptographicOperations.FixedTimeEquals(SHA256.HashData(image), raw.AsSpan(digestStart, 32)))
            throw new InvalidDataException("Image digest mismatch");
        return (raw[HeaderBytes..digestStart], image, key);
    }
}
